### PR TITLE
sql: replicating JSON empty array ordering found in Postgres

### LINKED ISSUE #105668: sql: decide on JSON ordering before 23.2 release
https://github.com/cockroachdb/cockroach/pull/101260 was created for the following reason:

> Currently, #97928 and #99275 are responsible for laying out
> a lexicographical ordering for JSON columns to be forward
> indexable in nature. This ordering is based on the rules
> posted by Postgres and is defined here: #99849
> 
> However, Postgres currently sorts the empty JSON array
> before any other JSON values. An issue has been opened:
> postgresql.org/message-id/17873-826fdc8bbcace4f1%40postgresql.org
> 
> Thus, this PR intends on replicating this behaviour until
> the issue has been identified and resolved by Postgres.

However, @mgartner says:

> Before this is merged, we'd like to check the SQL 2023 spec which reportedly defines the ordering of JSON values. If the ordering in the spec does not follow this empty array inconsistency of Postgres, we're probably fine with sticking with the current ordering. If the spec defines an ordering completely different than Postgres, we'll have to reconsider our options. This needs to be done before v23.2 is shipped.

This issue tracks that decision, and is a release blocker, since it will be a headache to change the ordering after the release goes out.

Epic: CRDB-24501

Jira issue: CRDB-29144

### PR DESCRIPTION
**Pull request by Shivs11**
Currently, #97928 and #99275 are responsible for laying out a
lexicographical ordering for JSON columns to be forward indexable in
nature. This ordering is based on the rules posted by Postgres and is
in #99849.

However, Postgres currently sorts the empty JSON array before any other
JSON values. A Postgres bug report for this has been opened:
https://www.postgresql.org/message-id/17873-826fdc8bbcace4f1%40postgresql.org

This PR intends on replicating the Postgres behavior.

Fixes #105668

Epic: CRDB-24501

Release note: None

### CODE REVIEW

**Review by yuzefovich — COMMENTED** [review]
LGTM, I only have nits. I'll defer to Marcus for approval though.

> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101260#-:-NSnbYS22czlVGM7ryqX:b47j0ap)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @Shivs11)

___
*[`-- commits` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101260#-NSnabjfEb_-OKN_ffj5):*
nit: I'd omit the release note since now the behavior matches Postgres (in other words, we are "fixing" difference from PG that was introduced in previous PRs, but those PRs haven't been a part of any release, so we don't need a "bug fix" release note).
___
*[`pkg/util/encoding/encoding.go` line 111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101260#-NSnavnO7ay57opa8rui:-NSnavnO7ay57opa8ruj:bjobupi) ([raw file](https://github.com/cockroachdb/cockroach/blob/27563ce86f4d77a2d0a5ee0230ccf009cbdf6571/pkg/util/encoding/encoding.go#L111)):*
> ```Go
> 
> 	// Postgres currently has a special case (maybe a bug) where
> 	// the empty JSON Array sorts before all other JSON values.
> ```

nit: consider putting a link to the PG bug report here too.
___
*[`pkg/util/encoding/encoding.go` line 3477 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101260#-NSnbPL2DuuCSKvAcjGM:-NSnbPL2DuuCSKvAcjGN:bb1e4xc) ([raw file](https://github.com/cockroachdb/cockroach/blob/27563ce86f4d77a2d0a5ee0230ccf009cbdf6571/pkg/util/encoding/encoding.go#L3477)):*
> ```Go
> // EncodeJSONArrayKeyMarker adds a JSON Array key encoding marker
> // to buf and returns the new buffer.
> func EncodeJSONArrayKeyMarker(buf []byte, dir Direction, length int64) []byte {
> ```

nit: perhaps `s/length/arrayLength/g` to be more explicit.

**Comment by mgartner** [comment]
Before this is merged, we'd like to check the SQL 2023 spec which reportedly defines the ordering of JSON values. If the ordering in the spec does not follow this empty array inconsistency of Postgres, we're probably fine with sticking with the current ordering. If the spec defines an ordering completely different than Postgres, we'll have to reconsider our options. This needs to be done before v23.2 is shipped.

**Review by mgartner — CHANGES_REQUESTED** [review]
> Reviewed 1 of 6 files at r1, 2 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101260)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Shivs11 and @yuzefovich)

___
*[`-- commits` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101260#-NSnabjfEb_-OKN_ffj5):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I'd omit the release note since now the behavior matches Postgres (in other words, we are "fixing" difference from PG that was introduced in previous PRs, but those PRs haven't been a part of any release, so we don't need a "bug fix" release note).
</blockquote></details>

Done.
___
*[`pkg/util/encoding/encoding.go` line 111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101260#-NSnavnO7ay57opa8rui:-Nb0yIENCx3ipDC2Q2tD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/27563ce86f4d77a2d0a5ee0230ccf009cbdf6571/pkg/util/encoding/encoding.go#L111)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: consider putting a link to the PG bug report here too.
</blockquote></details>

Done.
___
*[`pkg/util/encoding/encoding.go` line 3477 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101260#-NSnbPL2DuuCSKvAcjGM:-Nb0ySoQ6eaBjpDnoVHk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/27563ce86f4d77a2d0a5ee0230ccf009cbdf6571/pkg/util/encoding/encoding.go#L3477)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: perhaps `s/length/arrayLength/g` to be more explicit.
</blockquote></details>

Done.
___
*[`pkg/util/json/encoded.go` line 613 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101260#-Nb0z45_AC2RVYhRMA-h:-Nb0z45_AC2RVYhRMA-i:b5kyj8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e48064337fd6a405335e44f87c7842f27b5ffce/pkg/util/json/encoded.go#L613)):*
> ```Go
> 		return -1, nil
> 	}
> 	jDecoded, _ := j.decode()
> ```

Is it safe to ignore the error here? Probably not... I think we shouldn't decode the whole value either, just the marker to see if the array is empty.

**Review by mgartner — APPROVED** [review]
:lgtm: 

@yuzefovich do you mind giving this another quick look? I did some refactoring.

> Reviewed 2 of 6 files at r2, 1 of 3 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101260#-:-Nb184uaCcLQJ64puPWh:bkt91nn)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @Shivs11)

___
*[`pkg/util/json/encoded.go` line 613 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101260#-Nb0z45_AC2RVYhRMA-h:-Nb1563v43OIuJAgkQcP:blwxker) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e48064337fd6a405335e44f87c7842f27b5ffce/pkg/util/json/encoded.go#L613)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Is it safe to ignore the error here? Probably not... I think we shouldn't decode the whole value either, just the marker to see if the array is empty.
</blockquote></details>

Oh, we don't even need to decode in these cases. Cool. Done.

**Review by yuzefovich — APPROVED** [review]
Thanks! :lgtm:

Also perhaps consider updating the tech note too.

> Reviewed 2 of 6 files at r2, 2 of 3 files at r3, 2 of 2 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101260#-:-Nb1C1wsB5_uwGywlFAc:b-pgcpvd)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @Shivs11)

___
*[`-- commits` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101260#-NSnabjfEb_-OKN_ffj5):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

nit: also the PR description needs an update.
___
*[`pkg/util/json/json_test.go` line 80 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/101260#-Nb1BxvK53l1GIpf9xzh:-Nb1BxvK53l1GIpf9xzi:bert3ak) ([raw file](https://github.com/cockroachdb/cockroach/blob/8696c526ae87f8a40d284629ebb0afb996a7ca1b/pkg/util/json/json_test.go#L80)):*
> ```Go
> 		`false`,
> 		`true`,
> 		// In Postgres's sorting rules, the empty array comes before everything (even null),
> ```

nit: this comment should be updated.

**Review by mgartner — COMMENTED** [review]
> Also perhaps consider updating the tech note too.

Done.

> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101260#-:-Nb1Fp6e7Y-17gUNO3s9:bih2mwj)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @Shivs11)

___
*[`-- commits` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101260#-NSnabjfEb_-OKN_ffj5):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: also the PR description needs an update.
</blockquote></details>

Done.

**Comment by mgartner** [comment]
TFTR!

bors r+

**Pull request merged**
